### ISSUE TITLE
Spring Cloud AWS - SQS unable to pool multiple times to receive more than 10 messages' batch in a listener method.

### ISSUE THREAD

**Issue by shehrooz-10p** [awspring/spring-cloud-aws#1017]
**Type**: Bug

**Component**:
"SQS" 

**Describe the bug**
I am having trouble with batch processing using Spring Cloud for AWS (3.1.0) - SQS integration. it is stated in the documentation..

https://docs.awspring.io/spring-cloud-aws/docs/3.0.0/reference/html/index.html#batch-processing

"If a value greater than 10 is provided for maxMessagesPerPoll, the result of multiple polls will be combined and up to the respective amount of messages will be served to the listener."

But I am unable to receive more than 10 messages in my listeners even though maxMessagesPerPoll is set to 15 messages.

**Sample**
Following is the configurations that I am using for SQS.

@Configuration
public class SqsConfigs {

    @Autowired
    SqsAsyncClient sqsAsyncClient;

    @Bean
    public SqsTemplate sqsTemplate(SqsAsyncClient sqsAsyncClient) {
        return (SqsTemplate) SqsTemplate.newSyncTemplate(sqsAsyncClient);
    }

    @Bean
    SqsMessageListenerContainerFactory<Object> defaultSqsListenerContainerFactory(SqsAsyncClient sqsAsyncClient) {
        return SqsMessageListenerContainerFactory
                .builder()
                .configure(options -> options
                        .listenerMode(ListenerMode.BATCH)
                        .maxMessagesPerPoll(15)
                        .maxConcurrentMessages(20)
                        .pollTimeout(Duration.ofSeconds(20)))
                .sqsAsyncClient(sqsAsyncClient)
                .build();
    }
    
}

In my Listener, I have simplified the code to just print the messages as follows:.

 @SqsListener("TimeseriesSQS")
    public void listen(List<String> message) {
        log.info(message.toString());
    }
    
I have tried many configurations, but I am unable to receive more than 10 messages in my listener.

Is this functionality no more part of the Spring Cloud SQS or am I missing something here?

@mokamoto12 @jgslima, and @tomazfernandes

**Comment by tomazfernandes** [comment]
Hey @shehrooz-10p, this is unexpected. I'll take a look and get back to you.

I imagine you already tried that, but you can also set these values in the annotation itself.

**Comment by shehrooz-10p** [comment]
@tomazfernandes Thank you for the quick response.

Yeah, I have already tried it but unfortunately, the same results.

My understanding is that if I set maxMessagesPerPoll=15, then I should always receive 15 messages in my listener as a batch. If there are fewer messages, then it keeps on pooling multiple times until 15 messages are received.

**trace**

2024-01-11T23:17:12.386+05:00  INFO 37748 --- [           main] c.d.p.DataProcessingApplication          : Starting DataProcessingApplication using Java 17.0.7 with PID 37748 (D:\00 Projects\springSpace\DataProcessing\target\classes started by shehrooz.sattar in D:\00 Projects\springSpace\DataProcessing)
2024-01-11T23:17:12.388+05:00  INFO 37748 --- [           main] c.d.p.DataProcessingApplication          : No active profile set, falling back to 1 default profile: "default"
2024-01-11T23:17:15.324+05:00  INFO 37748 --- [ecycle-thread-1] a.c.s.l.AbstractMessageListenerContainer : Container io.awspring.cloud.sqs.sqsListenerEndpointContainer#0 started
2024-01-11T23:17:15.332+05:00  INFO 37748 --- [           main] c.d.p.c.StartupEventListener             : Handling context started event.
2024-01-11T23:17:16.637+05:00  INFO 37748 --- [ntContainer#0-1] c.d.p.c.SqsListenerForMessages           : [1, 2, 8, 3, 4, 5, 6, 9, 10, 7]
2024-01-11T23:17:16.910+05:00  INFO 37748 --- [ntContainer#0-2] c.d.p.c.SqsListenerForMessages           : [1, 2, 5, 6]
2024-01-11T23:17:17.177+05:00  INFO 37748 --- [ntContainer#0-3] c.d.p.c.SqsListenerForMessages           : [7, 8, 9, 10, 1, 2]
2024-01-11T23:17:17.455+05:00  INFO 37748 --- [ntContainer#0-4] c.d.p.c.SqsListenerForMessages           : [7, 8, 3, 4, 5, 6]
2024-01-11T23:17:17.740+05:00  INFO 37748 --- [ntContainer#0-5] c.d.p.c.SqsListenerForMessages           : [3, 4, 9, 10]

You can see in this trace sometimes I get 10 messages, other times even fewer. But my expectation was to receive 15 messages batch every time my listener is invoked.

**Comment by tomazfernandes** [comment]
It will wait for messages up to pollTimeout, after that it'll return the amount of messages that were available up to then.

So you'll necessarily receive "up to" the specified amount of messages, which is consistent with how the SQS API works.

We could add a configuration to block until the required amount of messages is available, but TBH I'm not convinced it's a good idea, and would likely conflict with AWS API e.g what would we do with the messages the API already returned?

The correct way to test this would be:
- Stop the app
- Send e.g. 200 messages or more to the queue
- Set concurrent messages to e.g. 50
- Set messages per poll to e.g. 50
- Set pollTimeout to 10 seconds (SQS max)
- Start the app

You should definitely see more than 10 messages at once in the listener.

Can you give that a try?

**Comment by shehrooz-10p** [comment]
Thank you for explaining it well. I believe the documentation is misleading to some extent and can be improved by stating that the messages will be returned on best best-effort basis up to the "maxMessagesPerPoll" and within the pollTimeout.
![image](https://github.com/awspring/spring-cloud-aws/assets/7940280/1dbf86f2-9269-4b13-90b2-57f0b2704b2a)


I agree with you the blocking of messages would not make sense and there is a chance that in some scenarios when there are not enough messages we'll be pooling for a very long time.

I was able to test it as per your recommendation, and I was able to receive around more or less 50 messages. The count is still inconsistent, but that is perfectly explainable because of the pollTimeout or not enough messages being available on the queue.

This resolves my issue. Thanks.
